### PR TITLE
[XLA:GPU} Add DynamicSliceFusionCmd command to command buffer

### PR DESCRIPTION
**Pull request by shawnwang18**

### CODE REVIEW

File: xla/service/gpu/runtime/command_buffer_cmd.cc [hunk 1]
```diff
@@ -1988,4 +1991,252 @@ CommandBufferCmd::BufferUsageVector CollectiveBroadcastCmd::buffers() {
...
+  // pre-compute host allocation requirements.
+  for (DynamicSliceThunk::SliceDef& slice : slices_) {
+    offsets_allocs_base_.push_back(offsets_allocs_size_);
+    if (slice.sliced_shape.has_value()) {
+      offsets_allocs_size_ += slice.sliced_shape->rank() * sizeof(int64_t);
+    }
+  }
+}
+
+bool DynamicSliceFusionCmd::force_update() {
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.cc:2038]: can you add documentation on how decition is made, it's quire hard to parse it from the codr

File: xla/service/gpu/runtime/command_buffer_cmd.cc [hunk 2]
```diff
@@ -1988,4 +1991,252 @@ CommandBufferCmd::BufferUsageVector CollectiveBroadcastCmd::buffers() {
...
+  return !llvm::all_of(slices_, [](DynamicSliceThunk::SliceDef& slice) {
+    if (slice.offsets.has_value()) {
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.cc:2040]:
```
if (!slice.offsets.has_valie()) return false;

return llvm::all_of(...)
```

to reduce nesting

File: xla/service/gpu/runtime/command_buffer_cmd.cc [hunk 3]
```diff
@@ -1988,4 +1991,252 @@ CommandBufferCmd::BufferUsageVector CollectiveBroadcastCmd::buffers() {
...
+      return llvm::all_of(slice.offsets.value(),
+                          [](DynamicSliceThunk::Offset offset) {
+                            if (std::holds_alternative<uint64_t>(offset))
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.cc:2043]: return std::holds_alternative

**Review by ezhulenev — APPROVED** [review]

File: xla/service/gpu/runtime/command_buffer_cmd.h [hunk 4]
```diff
@@ -1146,6 +1148,62 @@ class CollectiveBroadcastCmd : public CollectiveCmd {
...
+      std::unique_ptr<CommandBufferCmdSequence> embedded_commands,
+      std::vector<std::optional<BufferAllocation::Slice>> arguments,
+      std::vector<std::unique_ptr<BufferAllocation>> fake_allocations_,
+      std::vector<std::optional<std::vector<DynamicSliceThunk::Offset>>>
+          offsets,
+      std::vector<std::optional<Shape>> orig_shapes,
+      std::vector<std::optional<Shape>> sliced_shapes,
+      std::vector<std::optional<uint64_t>> offset_byte_sizes);
+
+  absl::Status Initialize(const Thunk::InitializeParams& params,
```
Comment (shraiysh) [inline xla/service/gpu/runtime/command_buffer_cmd.h:1168]: nit: @shawnwang18 I saw a warning about this function missing an `override` or `final` annotation.
